### PR TITLE
Test cancellation at every `await` during request handling

### PR DESCRIPTION
**Pull request by squahtx**
~~I'm throwing up this PR for discussion rather than to get it merged as-is.~~
~~The code's pretty magical and I'm not sure if it's something we want to merge.~~
EDIT: Consensus seems to be that we're happy to have integration tests, even if they're magical.
I'm putting this up for a real review now.

The interesting commit is "Add `_test_cancellation_at_every_await` helper method".

## Motivation
Once we start enabling cancellation for various endpoints, it would be desirable to have some sort of integration testing for it as part of CI.

Improper handling of cancellation can manifest as anything from logging contexts going missing to stuck requests to certain endpoints. As it's uncommon for clients to disconnect before a request finishes processing, any bugs won't be immediately obvious after deployment to production and will bite at random times.

## Testing Approach
It turns out that when you `await` a `Deferred`, Python calls `Deferred.__await__().__next__()`. `Deferred.__await__()` just returns `self`.

And `__next__` controls what happens to the `await`:
 * It can return a `Deferred`, which blocks the coroutine.
   Python returns the `Deferred` out of the `coroutine.send()` call which started the coroutine.
   https://github.com/twisted/twisted/blob/7d6387b2c0d6c8111277a96cc56c50efdff2a4bc/src/twisted/internet/defer.py#L1687
   The twisted coroutine-runner, `inlineCallbacks`, sees the return value and adds callbacks to resume the coroutine.
 * It can raise a `StopIteration` with the result of the await.
 * It can raise another exception to raise an exception out of the await.

We can patch `Deferred.__next__` to intercept `await`s and force them to block. Which gives us time to throw a cancellation into the request processing!

With that power in hand, we can construct a test that will run a request up to the first `await` (on a `Deferred`) and cancel the request. And then we can re-run the request, allow it to get a little further, then cancel it again once it hits the next `await`, and so on. ie. we can test a cancellation at every `await` on a `Deferred` along a single code path.

We identify when we see a new `await`s by looking at the stack. If it's a stack we've seen before, we let the `await` behave as normal.

This PR implements this approach.

## Logging
The test process is very magical and it's not clear what it's actually doing for a given endpoint. So I've added some debug logging at INFO level that will print a call tree for the endpoint under test, showing every `await` the test sees. There's an example below for the `GET /rooms/$room_id/members` endpoint. `PartialStateEventsTracker.await_full_state` shows up, as you might hope, and gets cancel-tested.

<details>
<summary>Logging from a run through `GET /rooms/$room_id/members`</summary>

Only the new parts of the call tree are printed for each request.
```
request-0  Running _test_cancellation_at_every_await for test_get_member_list_cancellation...
request-1  JsonResource wrapped_async_request_handler:275 (/home/squah/repos/synapse/synapse/http/server.py:275)
request-1    preserve_fn.<locals>.g:353 (/home/squah/repos/synapse/synapse/http/server.py:353)
request-1      JsonResource._async_render:559 (/home/squah/repos/synapse/synapse/http/server.py:559)
request-1        RoomMemberListRestServlet.on_GET:511 (/home/squah/repos/synapse/synapse/rest/client/room.py:511)
request-1          MessageHandler.get_state_events:225 (/home/squah/repos/synapse/synapse/handlers/message.py:225)
request-1            Auth.check_user_in_room_or_world_readable:693 (/home/squah/repos/synapse/synapse/api/auth.py:693)
request-1              Auth.check_user_in_room:109 (/home/squah/repos/synapse/synapse/api/auth.py:109)
request-1                > *force-blocked await* at StateHandler.get_current_state:175 (/home/squah/repos/synapse/synapse/state/__init__.py:175)
request-1  --- disconnected ---
request-2                StateHandler.get_current_state:179 (/home/squah/repos/synapse/synapse/state/__init__.py:179)
request-2                  StateHandler measured_func:106 (/home/squah/repos/synapse/synapse/util/metrics.py:106)
request-2                    StateHandler.resolve_state_groups_for_events:424 (/home/squah/repos/synapse/synapse/state/__init__.py:424)
request-2                      StateGroupStorage.get_state_groups_ids:587 (/home/squah/repos/synapse/synapse/storage/state.py:587)
request-2                        StateGroupStorage._get_state_group_for_events:806 (/home/squah/repos/synapse/synapse/storage/state.py:806)
request-2                          PartialStateEventsTracker.await_full_state:70 (/home/squah/repos/synapse/synapse/storage/util/partial_state_events_tracker.py:70)
request-2                            DeferredCacheListDescriptor wrapped:511 (/home/squah/repos/synapse/synapse/util/caches/descriptors.py:511)
request-2                              g:753 (/home/squah/repos/synapse/synapse/logging/context.py:753)
request-2                                run_in_background:814 (/home/squah/repos/synapse/synapse/logging/context.py:814)
request-2                                  EventsWorkerStore.get_partial_state_events:2001 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:2001)
request-2                                    DatabasePool.simple_select_many_batch:1719 (/home/squah/repos/synapse/synapse/storage/database.py:1719)
request-2                                      DatabasePool.runInteraction:796 (/home/squah/repos/synapse/synapse/storage/database.py:796)
request-2                                        delay_cancellation:750 (/home/squah/repos/synapse/synapse/util/async_helpers.py:750)
request-2                                          DatabasePool _runInteraction:768 (/home/squah/repos/synapse/synapse/storage/database.py:768)
request-2                                            > *blocking await* at DatabasePool.runWithConnection:891 (/home/squah/repos/synapse/synapse/storage/database.py:891)
request-2                                      > *blocking await* at DatabasePool.runInteraction:796 (/home/squah/repos/synapse/synapse/storage/database.py:796)
request-2                          > *blocking await* at PartialStateEventsTracker.await_full_state:70 (/home/squah/repos/synapse/synapse/storage/util/partial_state_events_tracker.py:70)
request-2  --- disconnected ---
request-3                        > *force-blocked await* at StateGroupStorage._get_state_group_for_events:808 (/home/squah/repos/synapse/synapse/storage/state.py:808)
request-3  --- disconnected ---
request-4                    StateHandler.resolve_state_groups_for_events:433 (/home/squah/repos/synapse/synapse/state/__init__.py:433)
request-4                      StateGroupStorage.get_state_group_delta:565 (/home/squah/repos/synapse/synapse/storage/state.py:565)
request-4                        DeferredCacheDescriptor _wrapped:351 (/home/squah/repos/synapse/synapse/util/caches/descriptors.py:351)
request-4                          g:753 (/home/squah/repos/synapse/synapse/logging/context.py:753)
request-4                            run_in_background:814 (/home/squah/repos/synapse/synapse/logging/context.py:814)
request-4                              StateGroupDataStore.get_state_group_delta:155 (/home/squah/repos/synapse/synapse/storage/databases/state/store.py:155)
request-4                                DatabasePool.runInteraction:796 (/home/squah/repos/synapse/synapse/storage/database.py:796)
request-4                                  delay_cancellation:750 (/home/squah/repos/synapse/synapse/util/async_helpers.py:750)
request-4                                    DatabasePool _runInteraction:768 (/home/squah/repos/synapse/synapse/storage/database.py:768)
request-4                                      > *blocking await* at DatabasePool.runWithConnection:891 (/home/squah/repos/synapse/synapse/storage/database.py:891)
request-4                                > *blocking await* at DatabasePool.runInteraction:796 (/home/squah/repos/synapse/synapse/storage/database.py:796)
request-4                      > *blocking await* at StateGroupStorage.get_state_group_delta:565 (/home/squah/repos/synapse/synapse/storage/state.py:565)
request-4  --- disconnected ---
request-5                StateHandler.get_current_state:186 (/home/squah/repos/synapse/synapse/state/__init__.py:186)
request-5                  EventsWorkerStore.get_event:388 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:388)
request-5                    EventsWorkerStore.get_events_as_list:480 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:480)
request-5                      EventsWorkerStore._get_events_from_cache_or_db:690 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:690)
request-5                        delay_cancellation:750 (/home/squah/repos/synapse/synapse/util/async_helpers.py:750)
request-5                          DataStore get_missing_events_from_db:670 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:670)
request-5                            EventsWorkerStore._get_events_from_db:985 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:985)
request-5                              EventsWorkerStore._enqueue_events:1135 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:1135)
request-5                                EventsWorkerStore._maybe_start_fetch_thread:817 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:817)
request-5                                  run_as_background_process:256 (/home/squah/repos/synapse/synapse/metrics/background_process_metrics.py:256)
request-5                                    run:243 (/home/squah/repos/synapse/synapse/metrics/background_process_metrics.py:243)
request-5                                      EventsWorkerStore._fetch_thread:823 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:823)
request-5                                        > *blocking await* at DatabasePool.runWithConnection:891 (/home/squah/repos/synapse/synapse/storage/database.py:891)
request-5                              > *blocking await* at EventsWorkerStore._enqueue_events:1139 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:1139)
request-5                      > *blocking await* at EventsWorkerStore._get_events_from_cache_or_db:690 (/home/squah/repos/synapse/synapse/storage/databases/main/events_worker.py:690)
request-5  --- disconnected ---
request-6          MessageHandler.get_state_events:230 (/home/squah/repos/synapse/synapse/handlers/message.py:230)
request-6            StateGroupWorkerStore.get_filtered_current_state_ids:268 (/home/squah/repos/synapse/synapse/storage/databases/main/state.py:268)
request-6              DatabasePool.runInteraction:796 (/home/squah/repos/synapse/synapse/storage/database.py:796)
request-6                delay_cancellation:750 (/home/squah/repos/synapse/synapse/util/async_helpers.py:750)
request-6                  DatabasePool _runInteraction:768 (/home/squah/repos/synapse/synapse/storage/database.py:768)
request-6                    > *blocking await* at DatabasePool.runWithConnection:891 (/home/squah/repos/synapse/synapse/storage/database.py:891)
request-6              > *blocking await* at DatabasePool.runInteraction:796 (/home/squah/repos/synapse/synapse/storage/database.py:796)
request-6  --- disconnected ---
request-7    preserve_fn.<locals>.g:357 (/home/squah/repos/synapse/synapse/http/server.py:357)
request-7      DirectServeJsonResource._send_response:426 (/home/squah/repos/synapse/synapse/http/server.py:426)
request-7        CallableMixin.__call__:1081 (/usr/lib/python3.8/unittest/mock.py:1081)
request-7          CallableMixin._mock_call:1085 (/usr/lib/python3.8/unittest/mock.py:1085)
request-7            CallableMixin._execute_mock_call:1155 (/usr/lib/python3.8/unittest/mock.py:1155)
request-7              respond_with_json:798 (/home/squah/repos/synapse/synapse/http/server.py:798)
request-7                run_in_background:814 (/home/squah/repos/synapse/synapse/logging/context.py:814)
request-7                  > *blocking await* at _async_write_json_to_request_in_thread:871 (/home/squah/repos/synapse/synapse/http/server.py:871)
```
</details>

## Limitations
* We can only test a single code path for a request, usually the happy path.
* We can't easily test code paths with contention, for example, in `ReadWriteLock`s.

## Pros of merging something like this
* Without this kind of test, we won't have any integration tests for cancellation.
* It's very easy to add a new test for an endpoint.
* The approach is robust to code changes deep within the call tree.
* It catches some bugs. For example, the test will fail if I revert #12183.

## Cons of merging something like this
* The code is complex and there is an unusual amount of magic involved, such as patching `Deferred.__next__` and inspecting the stack.
* It still misses bugs, such as when I revert #12199.
* It's not clear what's being tested, unless you turn on logging and manually inspect the test logs.

## Alternatives
* We could manually construct tests for each endpoint? They'll have less coverage than the approach in this PR.
* We could forgo integration tests entirely?
* We could patch `defer.inlineCallbacks` instead, like we do for `_check_yield_points`? We won't be able to force `await`s to block if we go for this.

### CODE REVIEW

**Comment by erikjohnston** [comment]
Wow, this is quite magical. Broadly, I think if this is the least-worst option for testing this stuff then we should merge it. In my mind awful testing strategies are better than *no* testing strategies? :shrug:

**Comment by reivilibre** [comment]
It is quite magical, but I can still make sense of it so it's not *too* bad. For a testing strategy I think this is reasonable and I wouldn't turn it down if the alternative is weaker or no coverage.
With this kind of helper, it's important to consider what the experience will be like for someone who ends up breaking one of these tests and has to figure out what's going on. I think the logging output is very helpful here so I don't think I have much to complain about.

I think it'd be better knowing about these problems rather than covering them up by not having such a testing scheme, so I'd say 'merge it'. If we really regret it for some reason later, we can just rip them out — they're only tests after all.

**Review by richvdh — COMMENTED** [review]
As others have said: it's magical, but not awful, and I don't think you're digging into anything that is likely to change in the near future.

It's definitely a net win.

**Comment by squahtx** [comment]
Thanks for your thoughts everyone! Sounds like we think the benefits outweigh the drawbacks, so I'll put this up for real review now.

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:192]: why do we need to patch this? Can we not just check the status of `channel` (in particular, check `channel.is_finished` instead of `respond_mock.called`)?

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:451]: I think this and the other methods below can be turned into regular functions. There's no need for it all to be part of the same object.

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:284]: Suggest making this whole thing an object, and turn `create_deferred___next___patch` into its constructor.

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:206]: I misunderstood this the first time I read it, and thought that `await_result=False` was a bad thing to do.

```suggestion
                                "ensure `await_result=False` is passed to `make_request`.",
```

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:202]: why would `request_number` be zero when it starts at 1?

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:218]: I'd inline this (and the other calls to `reactor.pump`) for clarity:

```suggestion
                            reactor.advance(0.0)
```

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:240]: can we just `return channel` and leave these up to the caller?

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:119]: rather than passing this in, can we just take a `*args, **kwargs` to pass through to `make_request` ?

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:116]: I feel like this should be called `make_request_with_cancellation_test` or something

File: tests/http/server/_base.py
Comment (richvdh) [inline tests/http/server/_base.py:241]: we can drop this `else`, since there's a `return` above.

**Review by richvdh — CHANGES_REQUESTED** [review]
I've made a start on looking at this and made some suggestions. Could you have a look at what I've said so far, and then I'll take a closer look?

File: tests/http/server/_base.py
Comment (squahtx) [reply tests/http/server/_base.py:240]: Possibly? There may be an unfortunate interaction where we block the `await`s in `_async_write_json_to_request_in_thread`.
I'll have to test it.

File: tests/http/server/_base.py
Comment (squahtx) [reply tests/http/server/_base.py:119]: I was thinking that we might want to test federation requests in future, which have their own `make_signed_federation_request` function
https://github.com/matrix-org/synapse/blob/9d8e380d2e8267129de921b9b926257c36417cd2/tests/unittest.py#L795

though there is nothing stopping us from defining a `make_signed_federation_request_with_cancellation_test` variant.

**Comment by squahtx** [comment]
Thanks for the feedback. I think I've addressed all of it now. Mind taking another look?

**Review by richvdh — COMMENTED** [review]
looks generally good. A few comments.

File: tests/http/server/_base.py
Comment (squahtx) [reply tests/http/server/_base.py:148]: Hm, that comment isn't really accurate.
`set_current_context` will raise an `AssertionError` when re-starting a finished log context. That happens within a deferred callback, so the `AssertionError` bubbles up through the request handling code. The exception will either get caught or exit a `with LoggingContext():` block, which will try to change the logging context away from the finished one and raise another `AssertionError`.

Comment (squahtx) [reply tests/http/server/_base.py:148]: I'll see if I can reword it.

File: tests/http/server/_base.py
Comment (squahtx) [reply tests/http/server/_base.py]: Thanks for the reminder!

**Comment by squahtx** [comment]
> The tests seem to have broken after #12872

It looks like a bug in the test.

Before #12872, when running `RoomsMemberListTestCase.test_get_member_list_cancellation`, we saw:
```
...
request-5  --- disconnected ---
request-6          MessageHandler.get_state_events:220 (/home/squah/repos/synapse/synapse/handlers/message.py:220)
request-6            StateGroupWorkerStore.get_filtered_current_state_ids:333 (/home/squah/repos/synapse/synapse/storage/databases/main/state.py:333)
request-6              DatabasePool.runInteraction:834 (/home/squah/repos/synapse/synapse/storage/database.py:834)
request-6                delay_cancellation:762 (/home/squah/repos/synapse/synapse/util/async_helpers.py:762)
request-6                  DatabasePool _runInteraction:806 (/home/squah/repos/synapse/synapse/storage/database.py:806)
request-6                    > *blocking await* at DatabasePool.runWithConnection:929 (/home/squah/repos/synapse/synapse/storage/database.py:929)
request-6              > *blocking await* at DatabasePool.runInteraction:834 (/home/squah/repos/synapse/synapse/storage/database.py:834)
request-6  --- disconnected ---
request-7    preserve_fn.<locals>.g:370 (/home/squah/repos/synapse/synapse/http/server.py:370)
request-7      DirectServeJsonResource._send_response:439 (/home/squah/repos/synapse/synapse/http/server.py:439)
request-7        CallableMixin.__call__:1081 (/usr/lib/python3.8/unittest/mock.py:1081)
request-7          CallableMixin._mock_call:1085 (/usr/lib/python3.8/unittest/mock.py:1085)
request-7            CallableMixin._execute_mock_call:1155 (/usr/lib/python3.8/unittest/mock.py:1155)
request-7              respond_with_json:811 (/home/squah/repos/synapse/synapse/http/server.py:811)
request-7                run_in_background:814 (/home/squah/repos/synapse/synapse/logging/context.py:814)
request-7                  > *blocking await* at _async_write_json_to_request_in_thread:884 (/home/squah/repos/synapse/synapse/http/server.py:884)
```

After #12872, we see:
```
...
request-5  --- disconnected ---
request-6          MessageHandler.get_state_events:220 (/home/squah/repos/synapse/synapse/handlers/message.py:220)
request-6            StateStorageController.get_current_state_ids:392 (/home/squah/repos/synapse/synapse/storage/controllers/state.py:392)
request-6              PartialCurrentStateTracker.await_full_state:162 (/home/squah/repos/synapse/synapse/storage/util/partial_state_events_tracker.py:162)
request-6                RoomWorkerStore.is_partial_state_room:1150 (/home/squah/repos/synapse/synapse/storage/databases/main/room.py:1150)
request-6                  DatabasePool.simple_select_one_onecol:1566 (/home/squah/repos/synapse/synapse/storage/database.py:1566)
request-6                    DatabasePool.runInteraction:834 (/home/squah/repos/synapse/synapse/storage/database.py:834)
request-6                      delay_cancellation:762 (/home/squah/repos/synapse/synapse/util/async_helpers.py:762)
request-6                        DatabasePool _runInteraction:806 (/home/squah/repos/synapse/synapse/storage/database.py:806)
request-6                          > *blocking await* at DatabasePool.runWithConnection:929 (/home/squah/repos/synapse/synapse/storage/database.py:929)
request-6                    > *blocking await* at DatabasePool.runInteraction:834 (/home/squah/repos/synapse/synapse/storage/database.py:834)
request-6  --- disconnected ---
request-7  <lambda>:414 (/home/squah/repos/synapse/tests/server.py:414)
request-7    DatabasePool _runInteraction:806 (/home/squah/repos/synapse/synapse/storage/database.py:806)
request-7      > *force-blocked await* at DatabasePool.runWithConnection:929 (/home/squah/repos/synapse/synapse/storage/database.py:929)
request-7  --- disconnected ---
```
and the test got stuck here, since the await we've forced to block sits within a `delay_cancellation`.
`synapse/tests/server.py:414` is the `callFromThread` trampoline that gets queued after a database transaction completes. What's strange is that it _does_ get called in both the before and after cases, but:
 * runs `Deferred.send` in the before case
 * and runs our `Deferred.__next__` patch in the after case

for reasons I cannot explain.

I've added code to unblock force-blocked awaits while waiting for a cancellation response.

**Review by richvdh — APPROVED** [review]
lgtm

**Comment by squahtx** [comment]
> Do you know which callback `callFromThread` is trying to run?

It's running `Deferred.callback`, to resume the `_runInteraction > runWithConnection` coroutine.

> ... which implies that in the before case, this is happening after our patch has been removed, implying some sort of timing change?

The patch is definitely still installed in both cases. If I add the following debug to `callFromThread`:

<details>
<summary>`callFromThread` debugging hack</summary>

```diff
--- a/tests/server.py
+++ b/tests/server.py
@@ -411,7 +411,18 @@ class ThreadedMemoryReactorClock(MemoryReactorClock):
         """
         Make the callback fire in the next reactor iteration.
         """
-        cb = lambda: callback(*args, **kwargs)
+        import types
+        def cb():
+            if isinstance(callback, types.MethodType) and callback.__func__ == Deferred.callback:
+                d = callback.__self__
+                gotResult_callbacks = d.callbacks[1]  # `d.callbacks[0]` is `_set_context_cb`
+                gotResult_callback, gotResult_errback = gotResult_callbacks
+                gotResult, _args, _kwargs = gotResult_callback
+                coroutine = gotResult.__closure__[1].cell_contents
+                import sys
+                sys.stdout.write(f"    callFromThread: resuming coroutine {coroutine}\n")
+                sys.stdout.write(f"        d.__next__ = {d.__next__.__qualname__}\n")
+            callback(*args, **kwargs)
         # it's not safe to call callLater() here, so we append the callback to a
```
</details>

I get:

Before #12872:
```
    callFromThread: resuming coroutine <coroutine object DatabasePool.runInteraction.<locals>._runInteraction at 0x7f4fab21edc0>
        d.__next__ = Deferred__next__Patch.patch.<locals>.Deferred___next__
```
After #12872:
```
    callFromThread: resuming coroutine <coroutine object DatabasePool.runInteraction.<locals>._runInteraction at 0x7f0f1b3c7240>
        d.__next__ = Deferred__next__Patch.patch.<locals>.Deferred___next__
```

**More logs with extra debugging:**
Before #12872:
```
request-6  --- disconnected ---
    callFromThread: resuming coroutine <coroutine object DatabasePool.runInteraction.<locals>._runInteraction at 0x7f4fab647d40>
        d.__next__ = Deferred.send
    Deferred.send() got called at /home/squah/repos/synapse/synapse/storage/database.py:929
request-7  --- make_request() ---
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/state/__init__.py:176
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/util/partial_state_events_tracker.py:70
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/state.py:868
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/state.py:608
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/database.py:929
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/database.py:834
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    callFromThread: resuming coroutine <coroutine object DatabasePool.runInteraction.<locals>._runInteraction at 0x7f4fab21edc0>
        d.__next__ = Deferred__next__Patch.patch.<locals>.Deferred___next__
    Deferred.send() got called at /home/squah/repos/synapse/synapse/storage/database.py:929
    Deferred.send() got called at /home/squah/repos/synapse/synapse/storage/database.py:834
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/http/server.py:884
request-7    preserve_fn.<locals>.g:370 (/home/squah/repos/synapse/synapse/http/server.py:370)
request-7      DirectServeJsonResource._send_response:439 (/home/squah/repos/synapse/synapse/http/server.py:439)
request-7        CallableMixin.__call__:1081 (/usr/lib/python3.8/unittest/mock.py:1081)
request-7          CallableMixin._mock_call:1085 (/usr/lib/python3.8/unittest/mock.py:1085)
request-7            CallableMixin._execute_mock_call:1155 (/usr/lib/python3.8/unittest/mock.py:1155)
request-7              respond_with_json:811 (/home/squah/repos/synapse/synapse/http/server.py:811)
request-7                run_in_background:814 (/home/squah/repos/synapse/synapse/logging/context.py:814)
request-7                  > *blocking await* at _async_write_json_to_request_in_thread:884 (/home/squah/repos/synapse/synapse/http/server.py:884)
```

After #12872:
```
request-6  --- disconnected ---
    callFromThread: resuming coroutine <coroutine object DatabasePool.runInteraction.<locals>._runInteraction at 0x7f0f1b3a9ec0>
        d.__next__ = Deferred.send
    Deferred.send() got called at /home/squah/repos/synapse/synapse/storage/database.py:929
request-7  --- make_request() ---
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/state/__init__.py:176
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/util/partial_state_events_tracker.py:71
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/controllers/state.py:340
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/controllers/state.py:80
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/database.py:929
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/database.py:834
    Deferred.send() got called at /home/squah/repos/synapse/tests/http/server/_base.py:379
    callFromThread: resuming coroutine <coroutine object DatabasePool.runInteraction.<locals>._runInteraction at 0x7f0f1b3c7240>
        d.__next__ = Deferred__next__Patch.patch.<locals>.Deferred___next__
    Deferred.__next__ patch got called at /home/squah/repos/synapse/synapse/storage/database.py:929
request-7  cb:425 (/home/squah/repos/synapse/tests/server.py:425)
request-7    DatabasePool _runInteraction:806 (/home/squah/repos/synapse/synapse/storage/database.py:806)
request-7      > *force-blocked await* at DatabasePool.runWithConnection:929 (/home/squah/repos/synapse/synapse/storage/database.py:929)
```

I genuinely cannot see a difference between the before/after `callFromThread` executions except whether `__next__` or `send` get called.

**Pull request merged**
